<a href="https://colab.research.google.com/github/guyfrancis/dat1001/blob/main/Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pandas

Pandas is a Python library that includes a useful datastructure type called 'dataframe' as well as associated code that help manipulate and analyze data.
Dataframes are good for storing and manipulating data in tabular form, with rows and columns.

Let's start by importing the pandas library.
We will then create a dictionary containing information about certain people and we will use pandas to convert it to a dataframe.

In [13]:
import pandas as pd
import numpy as np

my_dict = { "Name": ["Bob", "Charlie", "Elise"], "Age": [32, 47, 25], "Height (ft)": [5.9, 6.0, 5.2]}

df = pd.DataFrame(my_dict)

df.head()

Name  Age  Height (ft)
0      Bob   32          5.9
1  Charlie   47          6.0
2    Elise   25          5.2

## DataFrame Operations

Now that we have the data stored in a dataframe, we can perform a number of operations.

You can reference a column by using the column header.



In [14]:
# Reference the column 'Age'
print(df['Age'])

0    32
1    47
2    25
Name: Age, dtype: int64


You can reference a particular table value with the **iloc** method, giving the numerical value of the row and column

In [15]:
# Print the entry in first row (row 0) and second column (column 1) of the data frame
print(df.iloc[0, 1])

32


In [16]:
# You can also use iloc to get a whole row
print(df.iloc[1])

Name           Charlie
Age                 47
Height (ft)        6.0
Name: 1, dtype: object


In [17]:
# The describe method gives some summary information about the data
df.describe()

Age  Height (ft)
count   3.000000      3.00000
mean   34.666667      5.70000
std    11.239810      0.43589
min    25.000000      5.20000
25%    28.500000      5.55000
50%    32.000000      5.90000
75%    39.500000      5.95000
max    47.000000      6.00000

In [18]:
# The shape property tells you how many rows and columns are in the data
df.shape

(3, 3)

In [19]:
# We can also get the number of rows and columns from the shape property
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of rows: 3
Number of columns: 3


# Modifying data
We can add or remove rows and columns and modify the data in particular locations. We can also make changes based on conditions.

We'll do the following with this example data:

1) We'll add a column called "Over 40" and we will set it to 'Yes' if the person is over 40.
2) We'll change the title of the 'name' column to 'first name'
3) We will change Bob's height to 5.8 ft


In [20]:
# Create a new column called 'Over 40' with default value '-'
df['Over 40']='-'
df

Name  Age  Height (ft) Over 40
0      Bob   32          5.9       -
1  Charlie   47          6.0       -
2    Elise   25          5.2       -

In [21]:
# Rename 'Name' column to 'First Name'
df.rename(columns={'Name':"First Name"})

First Name  Age  Height (ft) Over 40
0        Bob   32          5.9       -
1    Charlie   47          6.0       -
2      Elise   25          5.2       -

In [22]:
# Change Bob's height to 5.8
# His height is in row 0 column 2
# We will use the iloc(row, col) method
df.iloc[0, 2] = 5.8
df

Name  Age  Height (ft) Over 40
0      Bob   32          5.8       -
1  Charlie   47          6.0       -
2    Elise   25          5.2       -

In [24]:
# Finally we will update the 'Over 40' column so that it says 'Yes' if the person is over 40 and 'No' otherwise.
df['Over 40'] = np.where(df['Age'] > 40, 'Yes', 'No')
df

Name  Age  Height (ft) Over 40
0      Bob   32          5.8      No
1  Charlie   47          6.0     Yes
2    Elise   25          5.2      No

# Reading Data from a Webpage

We can also use pandas to grab tabular data from webpages. To do this, we first have to import the **requests** library so we can make an http request to grab a webpage.

Before we make the web request, navigate to the webpage by clicking on the link below so that you can see what the webpage looks like.

https://worldpopulationreview.com/country-rankings/coffee-producing-countries

You should see a page that includes a table of countries with their coffee production.

Before we make the http request, we need to set the http header to include information about where the request is coming from. Websites may otherwise block requests.



In [58]:
# Import the requests library for making http requests
import requests as r
# Create a header that says the request is coming form a browser-like agent.
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36',
}

In [59]:
# Make an http request to get the webpage with the following url
url = "https://worldpopulationreview.com/country-rankings/coffee-producing-countries"
page = r.get(url, headers = headers)

In [60]:
# Check that the request was successful. If so, the status code should be 200.
page.status_code

200

In [61]:
# Print the first few hundred characters of the webpage
# The actual content of the page, which is HTML, is stored in the 'content' attribute of the page object
print(page.content[0:200])

b'<!DOCTYPE html><html lang="en"> <head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1"><title>Coffee Producing Countries 2025</title><meta name="description" c'


Now we will use the pandas.read_html() method

In [62]:
# read_html() will find all the tables in the webpage and put them in a list.
tables = pd.read_html(page.content)
# In this case, there is only one table, so the length of the list should be 1.
len(tables)

1

Each table in the HTML page is converted into a dataframe in the tables list. As there is only one table, let's store it in a separate variable called cd (for coffee data).

In [63]:
# Store the one table as a dataframe in the variable cd
cd = tables[0]
# Check the type of this variable. It should say 'pandas.core.frame.DataFrame'
type(cd)

pandas.core.frame.DataFrame

In [64]:
# Let's look at the table
cd

Unnamed: 0                           Country Coffee Production 2022 (t)↓  \
0          NaN                            Brazil                        3.2M   
1          NaN                           Vietnam                          2M   
2          NaN                         Indonesia                      794.8K   
3          NaN                          Colombia                        665K   
4          NaN                          Ethiopia                      496.2K   
5          NaN                            Uganda                      393.9K   
6          NaN                              Peru                      352.6K   
7          NaN                             India                      338.6K   
8          NaN                          Honduras                      315.5K   
9          NaN          Central African Republic                      306.9K   
10         NaN                            Guinea                      261.6K   
11         NaN                         Guatemala                      225.5K   
12         NaN                            Mexico                      181.7K   
13         NaN                              Laos                        171K   
14         NaN                         Nicaragua                      170.2K   
15         NaN                             China                      108.9K   
16         NaN                        Costa Rica                       79.2K   
17         NaN                       Ivory Coast                         70K   
18         NaN                          Tanzania                       67.2K   
19         NaN                          DR Congo                       58.8K   
20         NaN                       Philippines                       58.3K   
21         NaN                         Venezuela                       57.2K   
22         NaN                             Kenya                       51.9K   
23         NaN                        Madagascar                       50.3K   
24         NaN                  Papua New Guinea                       40.8K   
25         NaN                             Yemen                       36.5K   
26         NaN                       El Salvador                       30.7K   
27         NaN                          Cameroon                       26.4K   
28         NaN                              Togo                       25.1K   
29         NaN                           Bolivia                       23.2K   
30         NaN                Dominican Republic                       20.6K   
31         NaN                            Rwanda                       18.9K   
32         NaN                          Thailand                       18.7K   
33         NaN                           Burundi                       17.2K   
34         NaN                            Angola                       16.8K   
35         NaN                       Timor-Leste                       11.7K   
36         NaN                            Malawi                         11K   
37         NaN                           Myanmar                        8.9K   
38         NaN                            Zambia                        8.6K   
39         NaN                           Jamaica                        8.6K   
40         NaN                           Ecuador                          8K   
41         NaN                            Panama                        7.5K   
42         NaN                              Cuba                        6.3K   
43         NaN                             Haiti                        5.1K   
44         NaN                         Sri Lanka                        4.8K   
45         NaN                 Equatorial Guinea                        4.1K   
46         NaN                          Malaysia                        3.8K   
47         NaN             Republic of the Congo                        3.1K   
48         NaN                      Sierra Leone                        2.6K   
49  

## Data Wrangling

We are now going to do some data 'wrangling' on the dataframe. Basically, we are going to fix some issues with the data as it present to make it easier to analyze.

You may notice there are some issues with these data:

1. There is a column full of 'NaN' which we should delete
2. Some of the column titles contain special characters which we don't need - we can rename the columns.
3. The coffee production column has a mixture of numbers and letters 3.2M means 3.2 million (tonnes) 794.8K mean 794.8 thousand (tonnes). We need to replace these with consistent numbers.

Firstly, we can change the column names.




In [65]:
# Rename the columns
cd.columns = ['Not used', 'Country', '2022 Production (tonnes)', '2022 Yield (kg/hectare)']
cd

Not used                           Country 2022 Production (tonnes)  \
0        NaN                            Brazil                     3.2M   
1        NaN                           Vietnam                       2M   
2        NaN                         Indonesia                   794.8K   
3        NaN                          Colombia                     665K   
4        NaN                          Ethiopia                   496.2K   
5        NaN                            Uganda                   393.9K   
6        NaN                              Peru                   352.6K   
7        NaN                             India                   338.6K   
8        NaN                          Honduras                   315.5K   
9        NaN          Central African Republic                   306.9K   
10       NaN                            Guinea                   261.6K   
11       NaN                         Guatemala                   225.5K   
12       NaN                            Mexico                   181.7K   
13       NaN                              Laos                     171K   
14       NaN                         Nicaragua                   170.2K   
15       NaN                             China                   108.9K   
16       NaN                        Costa Rica                    79.2K   
17       NaN                       Ivory Coast                      70K   
18       NaN                          Tanzania                    67.2K   
19       NaN                          DR Congo                    58.8K   
20       NaN                       Philippines                    58.3K   
21       NaN                         Venezuela                    57.2K   
22       NaN                             Kenya                    51.9K   
23       NaN                        Madagascar                    50.3K   
24       NaN                  Papua New Guinea                    40.8K   
25       NaN                             Yemen                    36.5K   
26       NaN                       El Salvador                    30.7K   
27       NaN                          Cameroon                    26.4K   
28       NaN                              Togo                    25.1K   
29       NaN                           Bolivia                    23.2K   
30       NaN                Dominican Republic                    20.6K   
31       NaN                            Rwanda                    18.9K   
32       NaN                          Thailand                    18.7K   
33       NaN                           Burundi                    17.2K   
34       NaN                            Angola                    16.8K   
35       NaN                       Timor-Leste                    11.7K   
36       NaN                            Malawi                      11K   
37       NaN                           Myanmar                     8.9K   
38       NaN                            Zambia                     8.6K   
39       NaN                           Jamaica                     8.6K   
40       NaN                           Ecuador                       8K   
41       NaN                            Panama                     7.5K   
42       NaN                              Cuba                     6.3K   
43       NaN                             Haiti                     5.1K   
44       NaN                         Sri Lanka                     4.8K   
45       NaN                 Equatorial Guinea                     4.1K   
46       NaN                          Malaysia                     3.8K   
47       NaN             Republic of the Congo                     3.1K   
48       NaN                      Sierra Leone                     2.6K   
49       NaN                     United States                     2.4K   
50       NaN                           Nigeria                     1.9K   
51       NaN                            Taiwan                      886   
52       NaN                 

In [44]:
# Drop column 1
cd = cd.drop(columns=['col1'])

In [45]:
cd.shape

(79, 3)

In [46]:
pd.set_option('display.max_rows', 79)

In [47]:
cd

Country 2022 Production (tonnes)  \
0                             Brazil                     3.2M   
1                            Vietnam                       2M   
2                          Indonesia                   794.8K   
3                           Colombia                     665K   
4                           Ethiopia                   496.2K   
5                             Uganda                   393.9K   
6                               Peru                   352.6K   
7                              India                   338.6K   
8                           Honduras                   315.5K   
9           Central African Republic                   306.9K   
10                            Guinea                   261.6K   
11                         Guatemala                   225.5K   
12                            Mexico                   181.7K   
13                              Laos                     171K   
14                         Nicaragua                   170.2K   
15                             China                   108.9K   
16                        Costa Rica                    79.2K   
17                       Ivory Coast                      70K   
18                          Tanzania                    67.2K   
19                          DR Congo                    58.8K   
20                       Philippines                    58.3K   
21                         Venezuela                    57.2K   
22                             Kenya                    51.9K   
23                        Madagascar                    50.3K   
24                  Papua New Guinea                    40.8K   
25                             Yemen                    36.5K   
26                       El Salvador                    30.7K   
27                          Cameroon                    26.4K   
28                              Togo                    25.1K   
29                           Bolivia                    23.2K   
30                Dominican Republic                    20.6K   
31                            Rwanda                    18.9K   
32                          Thailand                    18.7K   
33                           Burundi                    17.2K   
34                            Angola                    16.8K   
35                       Timor-Leste                    11.7K   
36                            Malawi                      11K   
37                           Myanmar                     8.9K   
38                            Zambia                     8.6K   
39                           Jamaica                     8.6K   
40                           Ecuador                       8K   
41                            Panama                     7.5K   
42                              Cuba                     6.3K   
43                             Haiti                     5.1K   
44                         Sri Lanka                     4.8K   
45                 Equatorial Guinea                     4.1K   
46                          Malaysia                     3.8K   
47             Republic of the Congo                     3.1K   
48                      Sierra Leone                     2.6K   
49                     United States                     2.4K   
50                           Nigeria                     1.9K   
51                            Taiwan                      886   
52                       Puerto Rico                      843   
53                        Mozambique                      816   
54                             Ghana                      736   
55               Trinidad and Tobago                      681   
56                          Zimbabwe                      661   
57                           Liberia                      654   
58                            Guyana                      600   
59                          Cambodia                      374   
60                          Dominica                      370   

In [49]:
for val in cd['2022 Production (tonnes)']:
  print(type(val))

<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class